druid/custom/rich_input/rich_input.lua

```diff
@@ -30,37 +31,105 @@ local SCHEME = {
 	PLACEHOLDER = "placeholder_text",
 	INPUT = "input_text",
 	CURSOR = "cursor_node",
+	HIGHLIGHT = "highlight_node",
 }
 
 
+local function set_cursor(self)
+	if self.touch_pos_x then
+		local text = self.input:get_text()
+		local node_pos = gui.get_screen_position(self.highlight)
+		local touch_delta_x = self.touch_pos_x - node_pos.x
+		local letters_count = utf8.len(self.input:get_text())
+		local cursor_delta = 0
+		local gap = self.input.total_width/2 * -1
+
+		for i = 1, letters_count do
+			cursor_delta =  gap + self.text:get_text_size(string.sub(text, 1, i)) - self.half_cursor_width
+			if cursor_delta <= touch_delta_x then
+				gui.set_position(self.cursor, vmath.vector3(cursor_delta, 0, 0))
+				self.input.cursor_letter_index = i
+			end
+		end
+		
+		self.touch_pos_x = nil
+	else
+		local text = self.input:get_text()
+		local gap = self.input.total_width/2 * -1
+		local cursor_delta =  gap + self.text:get_text_size(string.sub(text, 1, self.input.cursor_letter_index)) - self.half_cursor_width
+		gui.set_position(self.cursor, vmath.vector3(cursor_delta, 0, 0))
+	end
+end
+
+
 local function animate_cursor(self)
 	gui.cancel_animation(self.cursor, gui.PROP_COLOR)
 	gui.set_color(self.cursor, vmath.vector4(1))
 	gui.animate(self.cursor, gui.PROP_COLOR, vmath.vector4(1,1,1,0), gui.EASING_INSINE, 0.8, 0, nil, gui.PLAYBACK_LOOP_PINGPONG)
 end
 
 
-local function update_text(self, text)
+local function update_text(self)
 	local text_width = self.input.total_width
+	local text_height = self.input.text_height
 	animate_cursor(self)
-	gui.set_position(self.cursor, vmath.vector3(text_width/2, 0, 0))
+	--gui.set_position(self.cursor, vmath.vector3(text_width/2, 0, 0))
 	gui.set_scale(self.cursor, self.input.text.scale)
+	gui.set_size(self.highlight, vmath.vector3(text_width, text_height, 0))
+	set_cursor(self)
+end
+
+
+local function clear_text(self, replace_with_symbol)
+	local spacer = replace_with_symbol or ""
+	self.input:set_text(spacer)
+	update_text(self)
+
+	if replace_with_symbol then
+		gui.set_position(self.cursor, vmath.vector3(self.input.total_width/2, 0, 0))
+		self.input.cursor_letter_index = 1
+	else
+		gui.set_position(self.cursor, vmath.vector3(0, 0, 0))
```

Comment:
You can create a `local TEMP_VECTOR = vector3(0)` in module scope, change their fields instead creating vector each time. It allows to save a bit of memory